material-components/material-components-web
# chore(infrastructure): Report screenshot test results on GitHub PR

### What it does

- Posts screenshot test statuses and diff report links on every PR
    - The "Details" link takes you to the Travis CI job A) while the test is running, and B) when it fails with an error
    - The "Details" link takes you to the report page when the screenshot test completes
    - Uses the [GItHub commit status API](https://developer.github.com/v3/repos/statuses/) and a bot account: @mdc-web-bot 
- Adds an encrypted credential file for the GitHub API
- Adds color and collapsible sections to Travis CI log output to make it easier to read

Notes:

- Travis CI screenshot tests currently diff against the PR's `golden.json` file, so the tests will pass if the browsers render what the PR says they do. It's still up to the PR author and reviewer to diff against `origin/master`'s `golden.json` and make sure the _change_ makes sense.
- We should create a follow up PR that links to a diff against `master`'s `golden.json` instead.

### How to test it

1. Create a PR that modifies a file in `packages/` or `test/screenshot/` (such as this one).
2. Scroll to the bottom of the PR in GItHub.
3. **After about 2 minutes**, a `screenshot-test/butter-bot` status should appear - _like magic_.

### Example output

#### Pending - Running screenshot tests:

![01 - pending](https://user-images.githubusercontent.com/409245/42746200-7d268222-888b-11e8-950c-fabac9c2a4f1.png)

#### Passed - No diffs:

* [Travis job](https://travis-ci.org/material-components/material-components-web/jobs/404316948)
* [Diff report](https://storage.googleapis.com/mdc-web-screenshot-tests/travis/2018/07/16/06_28_03_366/report/report.html)

![02 - passed](https://user-images.githubusercontent.com/409245/42746214-8c44b38c-888b-11e8-99cc-a60ad84c36cc.png)

#### Failed - 4 Diffs found:

* [Travis job](https://travis-ci.org/material-components/material-components-web/builds/404308287)
* [Diff report](https://storage.googleapis.com/mdc-web-screenshot-tests/travis/2018/07/16/05_54_55_095/report/report.html)

![03 - failed](https://user-images.githubusercontent.com/409245/42746226-a26afd9c-888b-11e8-8d50-59d338f1c0f4.png)

#### Full page:

![image](https://user-images.githubusercontent.com/409245/42746995-4b54db96-888f-11e8-9e72-974a80efe477.png)
